### ISSUE TITLE
[FEATURE] Improve reliability of structured outputs in AI Services

### ISSUE THREAD

**Issue by langchain4j** [langchain4j/langchain4j#859]
- [x] If LLM responds with a JSON wrapped in backticks and/or some additional text, try to extract that instead of failing
- [ ] Automatically enable [JSON mode](https://docs.langchain4j.dev/tutorials/ai-services#json-mode) if return type of AI Service is POJO and underlying LLM supports it
- [ ] (Configurable) Retry in case of parsing failure, feeding the error back to the LLM
- [ ] If return type is enum/boolean, automatically use `logitBias` if supported by the LLM
- [ ] Consider adding function calling as a second option for structured outputs
- [ ] Consider using existing libraries to generate/validate JSON schema

**Comment by ruizrube** [comment]
Hi,

The templates for parsing outputs were developed and tested using ChatGPT, isn't?. Open LLMs, such as Llama, are very more verbose in their answers.

kind regards. 
Thank you for such a stunning project.

**Comment by langchain4j** [comment]
@ruizrube yes, most of the high-level features were developed with at least `gpt-3.5-turbo`-level LLM in mind.

But now when we have llama3/mixtral/etc and [json mode](https://docs.langchain4j.dev/tutorials/ai-services#json-mode) in Ollama, it should work as good for most of the use cases.

But we need to implement this PR anyway to ensure that a long tail of corner cases is handled well.

**Comment by langchain4j** [comment]
Related: https://github.com/langchain4j/langchain4j/issues/1087

**Comment by langchain4j** [comment]
Related: https://github.com/langchain4j/langchain4j/discussions/786

**Comment by abbas-gadhia** [comment]
In case anyone is stuck with this, just enable the json mode of the LLM for the time being.

**Comment by mariofusco** [comment]
This has been solved (or at least largely mitigated) with the introduction of a [specific output guardrail](https://docs.quarkiverse.io/quarkus-langchain4j/dev/guardrails.html#_rewriting_the_llm_output) in the quarkus-langchain4j extension. I wonder if a similar approach can be ported or reimplemented directly into langchain4j.

**Comment by langchain4j** [comment]
@mariofusco yes, guardrails are expected to be backported from quarkus-lc4j to lc4j at some point, so they should definitely help in some cases.

Although I am not sure that "automatically enabling json mode or json schema" or "use tools for structured outputs" is something that fits into the guardrails concept well.

**Comment by mr-cloud** [comment]
seems like already support "If LLM responds with a JSON wrapped in backticks and/or some additional text, try to extract that instead of failing" in newer version like 1.9.0 and above. should we close this issue?

**Comment by mikehearn** [comment]
No, let's not close, there's still lots that can be done. It shouldn't be up to the user to learn about and use the JSON output guard rail or strict mode, for example (I just shipped a bug to some early users because I assumed this was done for me automatically).
